src/interp/interp.cc

```diff
@@ -1138,775 +1122,1263 @@ void Thread::Push(Ref ref) {
   values_.push_back(Value::Make(ref));
 }
 
-RunResult Thread::StepInternal(Trap::Ptr* out_trap) {
-  using O = Opcode;
+#define RETURN_IF(code)                        \
+  {                                            \
+    RunResult res = code;                      \
+    if (WABT_UNLIKELY(res != RunResult::Ok)) { \
+      return res;                              \
+    }                                          \
+    break;                                     \
+  }
 
-  u32& pc = frames_.back().offset;
-  auto& istream = mod_->desc().istream;
+RunResult Thread::Run(int num_instructions, Trap::Ptr* out_trap) {
+  using O = Opcode;
 
-  if (trace_stream_) {
-    istream.Trace(trace_stream_, pc, trace_source_.get());
-  }
+  DefinedFunc::Ptr func{store_, frames_.back().func};
+  for (; num_instructions > 0; --num_instructions) {
+    u32& pc = frames_.back().offset;
+    auto& istream = mod_->desc().istream;
 
-  auto instr = istream.Read(&pc);
-  switch (instr.op) {
```

Comment:
To avoid re-indenting this huge function (and to make this diff easier to read) pehaps keep the existing function as `RunOneInstruction()` and call it from within this loop?  if inclining is important it can be marked as such.

Also perhaps make a separate PR with the `using O = Opcode;` change and land that first?  (I thin that would also help make this diff more readable).

Reply:
Originally, interpreter was composed of one loop (`Thread::Run()`) that invokes a single opcode execution (`Thread::StepInternal()`) internally same as your suggestion. But this approach incurs exessive call overhead. So I just merge the loop and single execution call into one huge loop to reduce this call overhead (inlining the single opcode execution call into the loop). To do this merging, It seems that re-indenting is inevitable 😢 

`using O = Opcode;` is not newly added, but also included in the original single execution function (`Thread::StepInternal()`)
Shall I separate another PR for this?

Reply:
I see.

How about try marking the Step method as `__attribute__((always_inline))` ?

Reply:
I've updated the patch to avoid re-indenting as much as possible by adding a label (`NextOpcode`).
What do you think about this?

Actually, after this patch landed, I'm going to accelerate the interpreter performance more by adopting `threaded interpretation` method. This approach would greatly reduce switch-case overhead by directly jumping to the next opcode instead of decode-dispatch process. To apply this optimization, I first need to change the interpreter function as to one huge loop-switch code.

Reply:
I merged two `Thread::Run` functions and this showed a better performance for the same fibonacci code above.

* Update : 21.6 sec
* Origin : 25.1 sec